### PR TITLE
Fix #1368: Improve service delete error message

### LINKED ISSUE #1368: odo service create followed by odo service delete fails
[kind/bug]

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the chat and talk to us if you have a question rather than a bug or feature request.

The chat room is at: https://chat.openshift.io/developers/channels/odo

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?
- Operating System: Win10 Pro
- Output of `odo version`:

```
odo v0.0.19 (61f278a1)

Server: https://172.17.188.37:8443
OpenShift: v3.10.45
Kubernetes: v1.10.0+b81c8f8
```

## How did you run odo exactly?

`odo service create postgresql-persistent --plan default s1 && odo service delete s1 -f`

## Actual behavior

```
 ✓  Service 's1' was created                                                                                                                                                                   
Progress of the provisioning will not be reported and might take a long time.                                                                                                                  
You can see the current status by executing 'odo service list'                                                                                                                                 
                                                                                                                                                                                               
 ✗  unable to delete service because Service Catalog is not enabled in your cluster:                                                                                                           
unable to get the service list: no condition in status for {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:s1 GenerateName: Namespace:myproject SelfLink:/apis/servicecatalog.k8s.io/v1beta1/nam
espaces/myproject/serviceinstances/s1 UID:b86386ee-392c-11e9-ab64-0242ac120006 ResourceVersion:686 Generation:1 CreationTimestamp:2019-02-25 19:39:44 +0100 CET DeletionTimestamp:<nil> Deletio
nGracePeriodSeconds:<nil> Labels:map[app:0-0-19-oqjj app.kubernetes.io/component-name:s1 app.kubernetes.io/component-type:postgresql-persistent app.kubernetes.io/name:0-0-19-oqjj] Annotations
:map[] OwnerReferences:[] Initializers:nil Finalizers:[kubernetes-incubator/service-catalog] ClusterName:} Spec:{ClusterServiceClassExternalName:"postgresql-persistent", ClusterServicePlanExt
ernalName:"default"} Status:{Conditions:[] AsyncOpInProgress:false OrphanMitigationInProgress:false LastOperation:<nil> DashboardURL:<nil> CurrentOperation: ReconciledGeneration:0 ObservedGen
eration:0 OperationStartTime:<nil> InProgressProperties:<nil> ExternalProperties:<nil> ProvisionStatus: DeprovisionStatus:NotRequired}}                                                        
```

## Expected behavior
Either success or refusal but for another reason that Service Catalog is not enabled in your cluster

## Any logs, error output, etc?

### PR DESCRIPTION
**Pull request by metacosm**
## What is the purpose of this change? What does it change?
See $SUBJECT

## Was the change discussed in an issue?
fixes #1368 

## How to test changes?
Failure scenario in #1368 should now fail with a better error message.

### CODE REVIEW

**Comment by jeffmaury** [comment]
Does not work for me. I can't create the service. 

See the log:

```
$ odo service create postgresql-persistent --plan default s1 -v 10                                                                                                                             
I0228 11:39:24.983865    5140 loader.go:359] Config loaded from file C:\Users\JeffMAURY/.kube/config                                                                                           
I0228 11:39:25.021855    5140 occlient.go:440] Trying to connect to server 172.17.67.87:8443                                                                                                   
I0228 11:39:25.023867    5140 occlient.go:447] Server https://172.17.67.87:8443 is up                                                                                                          
I0228 11:39:25.024856    5140 round_trippers.go:386] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: odo/v0.0.0 (windows/amd64) kubernetes/$Format" -H "Authorization: Bea
rer hfe56khPhimCoRJ6a4Gd4xQRSPCuCseTFUFj13l31sk" 'https://172.17.67.87:8443/apis/user.openshift.io/v1/users/~'                                                                                 
I0228 11:39:25.037859    5140 round_trippers.go:405] GET https://172.17.67.87:8443/apis/user.openshift.io/v1/users/~ 200 OK in 12 milliseconds                                                 
I0228 11:39:25.037859    5140 round_trippers.go:411] Response Headers:                                                                                                                         
I0228 11:39:25.038860    5140 round_trippers.go:414]     Content-Type: application/json                                                                                                        
I0228 11:39:25.038860    5140 round_trippers.go:414]     Date: Thu, 28 Feb 2019 10:39:25 GMT                                                                                                   
I0228 11:39:25.039856    5140 round_trippers.go:414]     Content-Length: 306                                                                                                                   
I0228 11:39:25.039856    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.040855    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.041856    5140 request.go:897] Response Body: {"kind":"User","apiVersion":"user.openshift.io/v1","metadata":{"name":"developer","selfLink":"/apis/user.openshift.io/v1/users/dev
eloper","uid":"be782b69-3b27-11e9-a4ca-00155d00282a","resourceVersion":"1879","creationTimestamp":"2019-02-28T07:09:09Z"},"identities":["anypassword:developer"],"groups":null}                
I0228 11:39:25.043874    5140 occlient.go:370] isLoggedIn err:  <nil>                                                                                                                          
 output: "developer"                                                                                                                                                                           
I0228 11:39:25.044858    5140 round_trippers.go:386] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: odo/v0.0.0 (windows/amd64) kubernetes/$Format" -H "Authorization: Bea
rer hfe56khPhimCoRJ6a4Gd4xQRSPCuCseTFUFj13l31sk" 'https://172.17.67.87:8443/apis/project.openshift.io/v1/projects'                                                                             
I0228 11:39:25.046857    5140 round_trippers.go:405] GET https://172.17.67.87:8443/apis/project.openshift.io/v1/projects 200 OK in 1 milliseconds                                              
I0228 11:39:25.047855    5140 round_trippers.go:411] Response Headers:                                                                                                                         
I0228 11:39:25.047855    5140 round_trippers.go:414]     Content-Type: application/json                                                                                                        
I0228 11:39:25.047855    5140 round_trippers.go:414]     Date: Thu, 28 Feb 2019 10:39:25 GMT                                                                                                   
I0228 11:39:25.048857    5140 round_trippers.go:414]     Content-Length: 1284                                                                                                                  
I0228 11:39:25.048857    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.048857    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.049854    5140 request.go:897] Response Body: {"kind":"ProjectList","apiVersion":"project.openshift.io/v1","metadata":{"selfLink":"/apis/project.openshift.io/v1/projects"},"ite
ms":[{"metadata":{"name":"jeff","selfLink":"/apis/project.openshift.io/v1/projects/jeff","uid":"13ead853-3b3d-11e9-b414-00155d00282a","resourceVersion":"19456","creationTimestamp":"2019-02-28
T09:41:52Z","annotations":{"openshift.io/description":"","openshift.io/display-name":"","openshift.io/requester":"developer","openshift.io/sa.scc.mcs":"s0:c11,c10","openshift.io/sa.scc.supple
mental-groups":"1000130000/10000","openshift.io/sa.scc.uid-range":"1000130000/10000"}},"spec":{"finalizers":["openshift.io/origin","kubernetes"]},"status":{"phase":"Active"}},{"metadata":{"na
me":"myproject","selfLink":"/apis/project.openshift.io/v1/projects/myproject","uid":"be89afc1-3b27-11e9-b414-00155d00282a","resourceVersion":"1903","creationTimestamp":"2019-02-28T07:09:09Z",
"annotations":{"openshift.io/description":"Initial developer project","openshift.io/display-name":"My Project","openshift.io/requester":"developer","openshift.io/sa.scc.mcs":"s0:c11,c5","open
shift.io/sa.scc.supplemental-groups":"1000120000/10000","openshift.io/sa.scc.uid-range":"1000120000/10000"}},"spec":{"finalizers":["openshift.io/origin","kubernetes"]},"status":{"phase":"Acti
ve"}}]}                                                                                                                                                                                        
I0228 11:39:25.054859    5140 context.go:231] Ignoring error as it usually means flag wasn't set: flag accessed but not defined: component                                                     
I0228 11:39:25.054859    5140 round_trippers.go:386] curl -k -v -XGET  -H "User-Agent: odo/v0.0.0 (windows/amd64) kubernetes/$Format" -H "Accept: application/json, */*" -H "Authorization: Bea
rer hfe56khPhimCoRJ6a4Gd4xQRSPCuCseTFUFj13l31sk" 'https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/clusterserviceclasses?fieldSelector=spec.externalName%3Dpostgresql-persistent'  
I0228 11:39:25.065853    5140 round_trippers.go:405] GET https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/clusterserviceclasses?fieldSelector=spec.externalName%3Dpostgresql-persis
tent 200 OK in 7 milliseconds                                                                                                                                                                  
I0228 11:39:25.065853    5140 round_trippers.go:411] Response Headers:                                                                                                                         
I0228 11:39:25.066878    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.067855    5140 round_trippers.go:414]     Content-Type: application/json                                                                                                        
I0228 11:39:25.067855    5140 round_trippers.go:414]     Date: Thu, 28 Feb 2019 10:39:25 GMT                                                                                                   
I0228 11:39:25.067855    5140 round_trippers.go:414]     Content-Length: 1640                                                                                                                  
I0228 11:39:25.068858    5140 request.go:897] Response Body: {"kind":"ClusterServiceClassList","apiVersion":"servicecatalog.k8s.io/v1beta1","metadata":{"selfLink":"/apis/servicecatalog.k8s.io
/v1beta1/clusterserviceclasses","resourceVersion":"290"},"items":[{"metadata":{"name":"9737aca7-3b27-11e9-a4ca-00155d00282a","selfLink":"/apis/servicecatalog.k8s.io/v1beta1/clusterserviceclas
ses/9737aca7-3b27-11e9-a4ca-00155d00282a","uid":"cebc8fd7-3b3f-11e9-b118-0242ac120005","resourceVersion":"87","creationTimestamp":"2019-02-28T10:01:24Z"},"spec":{"externalName":"postgresql-pe
rsistent","externalID":"9737aca7-3b27-11e9-a4ca-00155d00282a","description":"PostgreSQL database service, with persistent storage. For more information about using this template, including Op
enShift considerations, see https://github.com/sclorg/postgresql-container/.\n\nNOTE: Scaling to more than one replica is not supported. You must have persistent volumes available in your clu
ster to use this template.","bindable":true,"bindingRetrievable":false,"planUpdatable":false,"externalMetadata":{"console.openshift.io/iconClass":"icon-postgresql","displayName":"PostgreSQL",
"documentationUrl":"https://docs.openshift.org/latest/using_images/db_images/postgresql.html","longDescription":"This template provides a standalone PostgreSQL server with a database created.
  The database is stored on persistent storage.  The database name, username, and password are chosen via parameters when provisioning this service.","providerDisplayName":"Red Hat, Inc.","su
pportUrl":"https://access.redhat.com"},"tags":["database","postgresql"],"clusterServiceBrokerName":"template-service-broker"},"status":{"removedFromBrokerCatalog":false}}]}                   
I0228 11:39:25.072857    5140 round_trippers.go:386] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: odo/v0.0.0 (windows/amd64) kubernetes/$Format" -H "Authorization: Bea
rer hfe56khPhimCoRJ6a4Gd4xQRSPCuCseTFUFj13l31sk" 'https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/clusterserviceplans/default'                                                    
I0228 11:39:25.078867    5140 round_trippers.go:405] GET https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/clusterserviceplans/default 404 Not Found in 4 milliseconds              
I0228 11:39:25.078867    5140 round_trippers.go:411] Response Headers:                                                                                                                         
I0228 11:39:25.079856    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.079856    5140 round_trippers.go:414]     Content-Type: application/json                                                                                                        
I0228 11:39:25.079856    5140 round_trippers.go:414]     Date: Thu, 28 Feb 2019 10:39:25 GMT                                                                                                   
I0228 11:39:25.080857    5140 round_trippers.go:414]     Content-Length: 266                                                                                                                   
I0228 11:39:25.080857    5140 request.go:897] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"clusterserviceplans.servicecatalog.k8s.io \"default
\" not found","reason":"NotFound","details":{"name":"default","group":"servicecatalog.k8s.io","kind":"clusterserviceplans"},"code":404}                                                        
I0228 11:39:25.081858    5140 round_trippers.go:386] curl -k -v -XGET  -H "User-Agent: odo/v0.0.0 (windows/amd64) kubernetes/$Format" -H "Authorization: Bearer hfe56khPhimCoRJ6a4Gd4xQRSPCuCse
TFUFj13l31sk" -H "Accept: application/json, */*" 'https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/namespaces/jeff/serviceinstances?fieldSelector=metadata.name%3Ds1&labelSelector=
app.kubernetes.io%2Fname%3Dtest-plmu'                                                                                                                                                          
I0228 11:39:25.086855    5140 round_trippers.go:405] GET https://172.17.67.87:8443/apis/servicecatalog.k8s.io/v1beta1/namespaces/jeff/serviceinstances?fieldSelector=metadata.name%3Ds1&labelSe
lector=app.kubernetes.io%2Fname%3Dtest-plmu 400 Bad Request in 2 milliseconds                                                                                                                  
I0228 11:39:25.086855    5140 round_trippers.go:411] Response Headers:                                                                                                                         
I0228 11:39:25.090872    5140 round_trippers.go:414]     Cache-Control: no-store                                                                                                               
I0228 11:39:25.092858    5140 round_trippers.go:414]     Content-Type: application/json                                                                                                        
I0228 11:39:25.094864    5140 round_trippers.go:414]     Date: Thu, 28 Feb 2019 10:39:25 GMT                                                                                                   
I0228 11:39:25.094864    5140 round_trippers.go:414]     Content-Length: 155                                                                                                                   
I0228 11:39:25.095874    5140 request.go:897] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"field label not supported: metadata.name","reason":
"BadRequest","code":400}                                                                                                                                                                       
I0228 11:39:25.096856    5140 cmdutils.go:21] Error:                                                                                                                                           
unable to list ServiceInstances: field label not supported: metadata.name                                                                                                                      
 ✗  field label not supported: metadata.name                                                                                                                                                   
```
And here is the output of `odo catalog list services`:

```
NAME                                              PLANS  
amq63-basic                                       default
amq63-persistent                                  default
amq63-persistent-ssl                              default
amq63-ssl                                         default
cakephp-mysql-persistent                          default
dancer-mysql-persistent                           default
datagrid65-mysql-persistent                       default
datagrid71-basic                                  default
datagrid71-mysql-persistent                       default
datagrid71-postgresql-persistent                  default
datagrid72-basic                                  default
datagrid72-mysql-persistent                       default
datagrid72-postgresql-persistent                  default
datavirt63-basic-s2i                              default
datavirt63-extensions-support-s2i                 default
datavirt63-secure-s2i                             default
decisionserver64-amq-s2i                          default
decisionserver64-basic-s2i                        default
decisionserver64-https-s2i                        default
django-psql-persistent                            default
eap70-amq-persistent-s2i                          default
eap70-basic-s2i                                   default
eap70-mysql-persistent-s2i                        default
eap70-postgresql-persistent-s2i                   default
eap70-tx-recovery-s2i                             default
eap71-amq-persistent-s2i                          default
eap71-basic-s2i                                   default
eap71-mysql-persistent-s2i                        default
eap71-postgresql-persistent-s2i                   default
eap71-tx-recovery-s2i                             default
fuse70-console                                    default
fuse70-console-cluster                            default
jenkins-ephemeral                                 default
jenkins-pipeline-example                          default
jws31-tomcat7-basic-s2i                           default
jws31-tomcat7-mongodb-persistent-s2i              default
jws31-tomcat7-mysql-persistent-s2i                default
jws31-tomcat7-postgresql-persistent-s2i           default
jws31-tomcat8-basic-s2i                           default
jws31-tomcat8-mongodb-persistent-s2i              default
jws31-tomcat8-mysql-persistent-s2i                default
jws31-tomcat8-postgresql-persistent-s2i           default
mariadb-persistent                                default
mongodb-persistent                                default
mysql-persistent                                  default
nodejs-mongo-persistent                           default
openjdk18-web-basic-s2i                           default
postgresql-persistent                             default
processserver64-amq-mysql-persistent-s2i          default
processserver64-amq-postgresql-persistent-s2i     default
processserver64-basic-s2i                         default
processserver64-mysql-persistent-s2i              default
processserver64-postgresql-persistent-s2i         default
rails-pgsql-persistent                            default
rhdm70-full-persistent                            default
rhdm70-kieserver                                  default
rhdm70-kieserver-basic-s2i                        default
rhdm70-kieserver-https-s2i                        default
rhpam70-authoring                                 default
rhpam70-authoring-ha                              default
rhpam70-kieserver-externaldb                      default
rhpam70-kieserver-mysql                           default
rhpam70-kieserver-postgresql                      default
rhpam70-prod                                      default
rhpam70-prod-immutable-kieserver                  default
rhpam70-prod-immutable-monitor                    default
rhpam70-sit                                       default
rhpam70-trial-ephemeral                           default
s2i-fuse70-eap-camel-amq                          default
s2i-fuse70-eap-camel-cdi                          default
s2i-fuse70-eap-camel-cxf-jaxrs                    default
s2i-fuse70-eap-camel-cxf-jaxws                    default
s2i-fuse70-eap-camel-jpa                          default
s2i-fuse70-karaf-camel-amq                        default
s2i-fuse70-karaf-camel-log                        default
s2i-fuse70-karaf-camel-rest-sql                   default
s2i-fuse70-karaf-cxf-rest                         default
s2i-fuse70-spring-boot-camel                      default
s2i-fuse70-spring-boot-camel-amq                  default
s2i-fuse70-spring-boot-camel-config               default
s2i-fuse70-spring-boot-camel-drools               default
s2i-fuse70-spring-boot-camel-infinispan           default
s2i-fuse70-spring-boot-camel-teiid                default
s2i-fuse70-spring-boot-camel-xml                  default
s2i-fuse70-spring-boot-cxf-jaxrs                  default
s2i-fuse70-spring-boot-cxf-jaxws                  default
sso72-x509-https                                  default
sso72-x509-mysql-persistent                       default
sso72-x509-postgresql-persistent                  default
```

**Comment by metacosm** [comment]
@jeffmaury I can't reproduce the issue… :( Not sure why you can't get service instances via their `metadata.name` field…

**Comment by jeffmaury** [comment]
@metacosm please not that I'm on CDK 3.6 ie OCP 3.10

**Comment by cdrage** [comment]
@metacosm I was able to use this with the `apb` templates:


```sh
github.com/redhat-developer/odo  pr_1374 ✔                                                                                                                                       2d
▶ odo service create dh-awx-apb --plan default                       
 ✓  Creating service
 ✓  Service 'dh-awx-apb' was created
Progress of the provisioning will not be reported and might take a long time.
You can see the current status by executing 'odo service list'
```

I'll try with postgresql-persistent when it comes up on Minishift...

**Comment by cdrage** [comment]
Tried again on Minishift and I wasn't able to replicate the issue:

```sh
github.com/redhat-developer/odo  pr_1374 ✔                                                                                                                                      2d  ⍉
▶ ./odo service create postgresql-persistent --plan default s2
 ✓  Creating service
 ✓  Service 's2' was created
Progress of the provisioning will not be reported and might take a long time.
You can see the current status by executing 'odo service list'

```

File: pkg/occlient/occlient.go [hunk 1]
```diff
@@ -5,8 +5,10 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	labels2 "github.com/redhat-developer/odo/pkg/application/labels"
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: nitpick, but applabels may be a better name than labels2

File: pkg/service/service.go [hunk 2]
```diff
@@ -186,10 +186,15 @@ func List(client *occlient.Client, applicationName string) ([]ServiceInfo, error
 	// Iterate through serviceInstanceList and add to service
 	for _, elem := range serviceInstanceList {
 		conditions := elem.Status.Conditions
+		var status string
 		if len(conditions) == 0 {
-			return nil, fmt.Errorf("no condition in status for %+v", elem)
+			glog.Warningf("no condition in status for %+v, marking it as Unknown", elem)
```
Comment (cdrage) [inline pkg/service/service.go]: I know I was away for 3 weeks on PTO, but have we started using glog.Warningf now instead of the debug / erroring out?

**Comment by amitkrout** [comment]
@metacosm Please add some expected output that would be really helpful to understand the issue fix from user prospective

File: pkg/service/service.go (hunk 2 above)
Comment (metacosm) [reply pkg/service/service.go]: I wanted to be able to surface the fact that there is something weird going on with the status without returning an error (which was actually causing the issue Jeff was seeing). I'm all hears for a better option if you have one.

**Comment by surajnarwade** [comment]
@jeffmaury , can you please try again & let us know ?

File: pkg/occlient/occlient.go [hunk 3]
```diff
@@ -5,8 +5,10 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	applabels "github.com/redhat-developer/odo/pkg/application/labels"
 	"io"
 	"io/ioutil"
+	"k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/labels"
```
Comment (kadel) [inline pkg/occlient/occlient.go]: this should be `k8s.io/apimachinery/pkg/labels`

**Pull request closed without merging**